### PR TITLE
Wip master

### PR DESCRIPTION
**Pull request by azat-co**
new PR CC @pletcher @vicfriedman @AnnJohn

### CODE REVIEW

**Comment by victhevenot** [comment]
comments for @azat-co 
- change header on line 3 from `overview` to `intro`
- this intro needs a LOT of work. in order for students to feel like node patterns (whatever those are) are important, just telling them they will learn them doesn't feel like its important. What problem do they solve? how difficult would it be without them? let's set up a problem, and then try to solve the problem with things the students already know - we want them to feel the pain, and then present them with an awesome solution to that problem, which is the thing we're trying to teach them. 
  - in the spirit  of our conversation this morning - i feel like i've commented on a good portion of the intros about the same thing. let's try to keep this in mind when going back and reviewing as well as moving forward with new content
  - is your sync error handling section supposed to be an intro? its confusing. if it is, nix the `overview` section entirely and rename `sync error handling` to `intro`
- on line 39 you tell students to create a fil `async-throw.js` and put code in it but the file exists and hte code is already there. chose one or the other. also - the readme is missing the `console.log('KA...')` at the end of the try/catch(which makes the explanation on like 66 hard to understand if you don't actually run `async-throw.js`
- line 63  `When you`setTimeout`with 0, it will schedule the callback at 0 milliseconds, but on the next tick (same as iteration or cycle) of the event loop` the sentence doesn't quite make sense. what about the next tick of the event loop?
- why are we console.logging? `KA...` what is that?
- let's make sure our examples are relevant and set up appropriately for our students to follow along. example on line 79 has no explanation for what we're trying to have our code do
- 136 feels like it needs its own sub-header
- jump to concurrency seems abrupt, is there a way to flow?
- all the examples feel really arbitrary and unrelated - it would be great to connect them all to one domain. it's hard to follow the concepts when the examples are jumping around

@pletcher on to you

File: README.md [hunk 1]
```diff
@@ -1 +1,218 @@
-# node-patterns-concurrency
\ No newline at end of file
+# Concurrency and Other Async Patterns
+
+## Overview
+
+Asynchronous code is more efficient than synchronous, but to write good async code you need to know at least a few Node patterns.
```
Comment (pletcher) [inline README.md:5]: I'm not sure what this would mean (stated so generally). The performance implications of async vs. sync code are difficult to compare, and I don't think that we want to bog students down with this discussion.

Certainly it's generally the case in Node.js that async code will have better performance characteristics than similar code executed synchronously -- but I think we should not frame the problem around performance when we're talking about error-handling.

File: README.md [hunk 2]
```diff
@@ -1 +1,218 @@
...
+Error: BOOM!
+    at null._onTimeout (/Users/azat/Documents/Code/learn-co/node-patterns-concurrency/async-throw.js:2:33)
+    at Timer.listOnTimeout (timers.js:92:15)
+```
+
+
+Note: When you `setTimeout` with 0, it will schedule the callback at 0 milliseconds, but on the next tick (same as iteration or cycle) of the event loop. For this reason, Node will finish with all the statements in this file first and only after that on the next cycle, it will fire up the callback.
+
+
+The "KA..." message is shown first, then the delayed callback throws an error which kills the program. When we don't have asynchronous code, `try/catch` works. However, with asynchronous code `try/catch` is useless unless you put it inside of the callback. Why is that?
```
Comment (pletcher) [inline README.md:66]: Where is this "KA..." message coming from? It's in the example file but not above, so this statement is confusing.

File: README.md [hunk 3]
```diff
@@ -1 +1,218 @@
...
+
+If we go deeper under the hood, callbacks are scheduled by the event loop. They lose context of the  environment in which they were defined (and lose `try/catch` which is around them). In a way, future code (callbacks) and current code (`try/catch`) become separated by the Node's non-blocking I/O and event loop mechanisms. 
+
+What should we do?
+
+## Async Error Handling
```
Comment (pletcher) [inline README.md:72]: I thought we covered this in the previous lesson (or one very close to this)? Why do we have a whole half of a lesson for async error handling?

File: README.md [hunk 4]
```diff
@@ -1 +1,218 @@
...
+```js
+module.exports = function(inputFile, customers, callback){
+  return function writeFile(error){
+    if (error) return callback(error)
+    return callback(null, customers)
+  }
+}
+```
+
+By using `if (error) return callback(error)` you will NOT increase the indentation which leads to callback hell. As you know, `return` will stop the execution and return the value making this function an expression. We don't really care about it being the expression. We use `return` to control the flow.
```
Comment (pletcher) [inline README.md:101]: As mentioned before, I think indentation is only an aesthetic concern when it comes to callback hell -- and indentation doesn't lead to callback hell on its own (especially indentation inside a block statement). I don't think there's any reason to mention callback hell here at all.

Comment (pletcher) [inline README.md:101]: Rather, the salient point is that the JS interpreter won't barf on `if (error) return callback(error)`, and that it's a matter of style whether one uses the one-line version or

``` js
if (error) {
  return callback(error)
}
```

File: README.md [hunk 5]
```diff
@@ -1 +1,218 @@
...
+
+If you don't have an error (it's falsy), just continue with your logic. At some point, you might want to return some data back to the original caller via the callback. Pass the data as the second argument, leaving the first for the error which is `null` if everything went fine.
+
+Because of the way closures work in JavaScript/Node (they have access to outer scope), you can invoke `callback` at any level when you have multiple levels of nested callbacks:
+
+```js
+var fs = require('fs')
+var inputFile = './customers.csv'
+
+module.exports = function(callback){
```
Comment (pletcher) [inline README.md:111]: This example is a little difficult to parse for demonstrating the availability of an error callback throughout this function's scope.

File: README.md [hunk 6]
```diff
@@ -1 +1,218 @@
...
+    } else {
+      console.error('The input CSV file is not found')
+    }
+  })
+}
+```
+
+When using an object which implements Event Emitters, listen to `error` event with `on('error', callback)`. Similarly, when you implement your own object with events, trigger error with `emit('error', error)`. As you remember, you can pass any data with `emit()`.
+
+When all hell breaks lose, listen to `process.on('uncaughtException', callback)`. Think about it as the last resort. Don't try to resume the execution at this point. There might be some lost states. Just log and exit gracefully.
```
Comment (pletcher) [inline README.md:138]: I think students could benefit from seeing an example of this in lieu of the example above (they'll already have a reasonably solid grasp of scope in JS)

File: README.md [hunk 7]
```diff
@@ -1 +1,218 @@
...
+This is much better. But, we still have a problem because we want just a single point of continuation, not 10. In other words, `processResponses` must be called just once. It's the continuation of your program. Right now it's called 10 times! No bueno. How about a simple counter? It will work!
+
+```js
+var j = 0
+var responses = {}
+for (var i = 1; i<=10; i++) {
+  makeRequest(i, function(response) {
+    j++
+    console.log(response)
+    responses[i] = (response)
```
Comment (pletcher) [inline README.md:201]: This is going to overwrite `responses[11]` every time.
